Contents

01. Importing Libraries
02. Importing Data Sets
03. Finding The Aggregated Mean
04. Creating A Loyalty Flag
05. Checking Spending Habits
06. Create a Spending Flag
07. Create an Order Frequency Flag
08. Exporting

# 01. Importing Libraries

In [1]:
# Importing libraries

import pandas as pd
import numpy as np
import os

# 02. Importing Data Sets

In [2]:
# Set path

path = r'/Users/danioprea/Desktop/Data Analytics/06-2023 Instacart Basket Analysis'

In [3]:
# Import orders_products_merged.pkl

ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

In [4]:
ords_prods_merge.shape

(32404859, 19)

In [5]:
ords_prods_merge.head(30)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   
15           1  Chocolate Sandwich Cookies        61             19     5.8   
16           1  Chocolate Sandwich Cookies        61             19     5.8   
17           1  Chocolate Sandwich Cookies        61             19     5.8   
18           1  Chocolate Sandwich Cookies        61             19     5.8   
19           1  Chocolate Sandwich Cookies        61             19     5.8   
20           1  Chocolate Sandwich Cookies        61             19     5.8   
21           1  Chocolate Sandwich Cookies        61             19     5.8   
22           1  Chocolate Sandwich Cookies        61             19     5.8   
23           1  Chocolate Sandwich Cookies        61             19     5.8   
24           1  Chocolate Sandwich Cookies        61             19     5.8   
25           1  Chocolate Sandwich Cookies        61             19     5.8   
26           1  Chocolate Sandwich Cookies        61             19     5.8   
27           1  Chocolate Sandwich Cookies        61             19     5.8   
28           1  Chocolate Sandwich Cookies        61             19     5.8   
29           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id eval_set  order_number  orders_day_of_week  \
0    3139998      138    prior            28                   6   
1    1977647      138    prior            30                   6   
2     389851      709    prior             2                   0   
3     652770      764    prior             1                   3   
4    1813452      764    prior             3                   4   
5    1701441      777    prior            16                   1   
6    1871483      825    prior             3                   2   
7    1290456      910    prior            12                   3   
8     369558     1052    prior            10                   1   
9     589712     1052    prior            15                   1   
10   1798115     1379    prior             1                   5   
11    497193     1480    prior            36                   3   
12   1727826     1480    prior            38                   4   
13   2694698     1480    prior            39                   2   
14   3171753     1494    prior             6                   3   
15    186761     1494    prior             7                   2   
16    317121     1494    prior             9                   5   
17    200390     1540    prior             4                   2   
18    374436     1540    prior             5                   1   
19   2920553     1540    prior             6                   1   
20   2813049     1540    prior             7                   2   
21   2365998     1540    prior             8              

# 03. Finding The Aggregated Mean

In [6]:
#Step 2 & 3

# Find the mean of the order_number column group by department_id

ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

Some of the means from the department_ids are pretty similar between the subset and the whole dataframe. For example, the mean of department_id 1 is 15.577 in the subset and 15.458 in the whole dataframe. However, some differ by a decent amount. For example, the mean of department_id 21 is 25.535 in the subset and 22.902 in the whole dataframe.

# 04. Creating A Loyalty Flag

In [7]:
# Step 4

# Create the max_order column

ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [8]:
ords_prods_merge.head(15)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id eval_set  order_number  orders_day_of_week  \
0    3139998      138    prior            28                   6   
1    1977647      138    prior            30                   6   
2     389851      709    prior             2                   0   
3     652770      764    prior             1                   3   
4    1813452      764    prior             3                   4   
5    1701441      777    prior            16                   1   
6    1871483      825    prior             3                   2   
7    1290456      910    prior            12                   3   
8     369558     1052    prior            10                   1   
9     589712     1052    prior            15                   1   
10   1798115     1379    prior             1                   5   
11    497193     1480    prior            36                   3   
12   1727826     1480    prior            38                   4   
13   2694698     1480    prior            39                   2   
14   3171753     1494    prior             6                   3   

    order_hour_of_day  days_since_prior_order  add_to_cart_order  reordered  \
0                  11                     3.0                  5          0   
1                  17                    20.0                  1          1   
2                  21                     6.0                 20          0   
3                  13                     NaN                 10          0   
4                  17                     9.0                 11          1   
5                   7                    26.0                  7          0   
6                  14                    30.0                  2          0   
7                  10                    30.0                  1          0   
8                  20                    19.0                  1          0   
9                  12                    15.0                  2          1   
10                 11                     NaN                  1          0   
11                 13                     7.0                  1          0   
12                 13                     7.0                  3          1   
13                 12                    12.0                  4          1   
14                 10                     2.0                  8          0   

   _merge    price_range_loc     busiest_day     busiest_days  \
0    both  Mid-range product  Regularly busy   Regularly busy   
1    both  Mid-range product  Regularly busy   Regularly busy   
2    both  Mid-range product     Busiest day     Busiest days   
3    both  Mid-range product  Regularly busy  Least busy days   
4    both  Mid-range product      Least busy  Least busy day

In [9]:
# Derive columns using the loc() function

ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [10]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [11]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [12]:
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

# 05. Checking Spending Habits

In [15]:
# Use the groupby function to display the mean, min and max prices for each loyalty flag

ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

It looks like Loyal Customers, on average, spend less than either New customers and Regular customers.

# 06. Create a Spending Flag

In [16]:
# Step 6

# Create the average_spending column

ords_prods_merge['average_spending'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [20]:
ords_prods_merge.head(50)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   
15           1  Chocolate Sandwich Cookies        61             19     5.8   
16           1  Chocolate Sandwich Cookies        61             19     5.8   
17           1  Chocolate Sandwich Cookies        61             19     5.8   
18           1  Chocolate Sandwich Cookies        61             19     5.8   
19           1  Chocolate Sandwich Cookies        61             19     5.8   
20           1  Chocolate Sandwich Cookies        61             19     5.8   
21           1  Chocolate Sandwich Cookies        61             19     5.8   
22           1  Chocolate Sandwich Cookies        61             19     5.8   
23           1  Chocolate Sandwich Cookies        61             19     5.8   
24           1  Chocolate Sandwich Cookies        61             19     5.8   
25           1  Chocolate Sandwich Cookies        61             19     5.8   
26           1  Chocolate Sandwich Cookies        61             19     5.8   
27           1  Chocolate Sandwich Cookies        61             19     5.8   
28           1  Chocolate Sandwich Cookies        61             19     5.8   
29           1  Chocolate Sandwich Cookies        61             19     5.8   
30           1  Chocolate Sandwich Cookies        61             19     5.8   
31           1  Chocolate Sandwich Cookies        61             19     5.8   
32           1  Chocolate Sandwich Cookies        61             19     5.8   
33           1  Chocolate Sandwich Cookies        61             19     5.8   
34           1  Chocolate Sandwich Cookies        61             19     5.8   
35           1  Chocolate Sandwich Cookies        61             19     5.8   
36           1  Chocolate Sandwich Cookies        61             19     5.8   
37           1  Chocolate Sandwich Cookies        61             19     5.8   
38           1  Chocolate Sandwich Cookies        61             19     5.8   
39           1  Chocolate Sandwich Cookies        61             19     5.8   
40           1  Chocolate Sandwich Cookies        61             19     5.8   
41           1  Chocolate Sandwich Cookies        61             19     5.8   
42           1  Chocolate Sandwich Cookies        61             19     5.8   
43           1  Chocolate Sandwich Cookies        61             19     5.8   
44           1  Chocolate Sandwich Cookies        61             19     5.8   
45           1  Chocolate Sandwich Cookies        61             19     5.8   
46           1  Chocolate Sandwich Cookies        61             19     5.8   
47           1  Chocolate Sandwich Cookies        61             19     5.8   
48           1  Chocolate Sandwich Cookies        61             19     5.8   
49           1  Chocolate Sandwich Cookies        61  

In [21]:
# Derive spending_flag columns using the loc() function

ords_prods_merge.loc[ords_prods_merge['average_spending'] < 10, 'spending_flag'] = 'Low spender'

In [22]:
ords_prods_merge.loc[ords_prods_merge['average_spending'] >= 10, 'spending_flag'] = 'High spender'

In [24]:
# Check the frequency of the spending_flag column

ords_prods_merge['spending_flag'].value_counts(dropna = False)

Low spender     31770614
High spender      634245
Name: spending_flag, dtype: int64

In [26]:
# Return the first 60 rows of the user_id, average_spending, spending_flag, and prices columns

ords_prods_merge[['user_id', 'average_spending', 'spending_flag', 'prices']].head(60)

user_id  average_spending spending_flag  prices
0       138          6.935811   Low spender     5.8
1       138          6.935811   Low spender     5.8
2       709          7.930208   Low spender     5.8
3       764          4.972414   Low spender     5.8
4       764          4.972414   Low spender     5.8
5       777          6.935398   Low spender     5.8
6       825          5.957576   Low spender     5.8
7       910          6.680000   Low spender     5.8
8      1052          7.162500   Low spender     5.8
9      1052          7.162500   Low spender     5.8
10     1379          5.813636   Low spender     5.8
11     1480          5.081633   Low spender     5.8
12     1480          5.081633   Low spender     5.8
13     1480          5.081633   Low spender     5.8
14     1494          6.803797   Low spender     5.8
15     1494          6.803797   Low spender     5.8
16     1494          6.803797   Low spender     5.8
17     1540          6.607122   Low spender     5.8
18     1540          6.607122   Low spender     5.8
19     1540          6.607122   Low spender     5.8
20     1540          6.607122   Low spender     5.8
21     1540          6.607122   Low spender     5.8
22     1540          6.607122   Low spender     5.8
23     1540          6.607122   Low spender     5.8
24     1540          6.607122   Low spender     5.8
25     1540          6.607122   Low spender     5.8
26     1540          6.607122   Low spender     5.8
27     1540          6.607122   Low spender     5.8
28     1540          6.607122   Low spender     5.8
29     1540          6.607122   Low spender     5.8
30     1540          6.607122   Low spender     5.8
31     1540          6.607122   Low spender     5.8
32     1540          6.607122   Low spender     5.8
33     1540          6.607122   Low spender     5.8
34     1598          7.123529   Low spender     5.8
35     1598          7.123529   Low spender     5.8
36     1598          7.123529   Low spender     5.8
37     1647          4.853521   Low spender     5.8
38     1647          4.853521   Low spender     5.8
39     1647          4.853521   Low spender     5.8
40     1647          4.853521   Low spender     5.8
41     1647          4.853521   Low spender     5.8
42     2833          7.377143   Low spender     5.8
43     2850          8.054153   Low spender     5.8
44     2857          8.353793   Low spender     5.8
45     3010          6.452800   Low spender     5.8
46     3029          7.960606   Low spender     5.8
47     3395          6.322222   Low spender     5.8
48     3904          7.511111   Low spender     5.8
49     4006          7.441100   Low spender     5.8
50     4006          7.441100   Low spender     5.8
51     4006          7.441100   Low spender     5.8
52     4006          7.441100   Low spender     5.8
53     4006          7.441100   Low spender     5.8
54     4122          7.598765   Low spender     5.8
55     4141          6.033333   Low spender     5.8
56     4343          7.393939   Low spender     5.8
57     4553          5.715789   Low spender     5.8
58     4845          7.229630   Low spender     5.8
59     6866          4.807143   Low spender     5.8

# 07. Create an Order Frequency Flag

In [27]:
# Step 7

# Create the median_days column

ords_prods_merge['median_days'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [32]:
pd.options.display.max_rows = None
pd.options.display.max_columns = None

ords_prods_merge.head(100)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   
15           1  Chocolate Sandwich Cookies        61             19     5.8   
16           1  Chocolate Sandwich Cookies        61             19     5.8   
17           1  Chocolate Sandwich Cookies        61             19     5.8   
18           1  Chocolate Sandwich Cookies        61             19     5.8   
19           1  Chocolate Sandwich Cookies        61             19     5.8   
20           1  Chocolate Sandwich Cookies        61             19     5.8   
21           1  Chocolate Sandwich Cookies        61             19     5.8   
22           1  Chocolate Sandwich Cookies        61             19     5.8   
23           1  Chocolate Sandwich Cookies        61             19     5.8   
24           1  Chocolate Sandwich Cookies        61             19     5.8   
25           1  Chocolate Sandwich Cookies        61             19     5.8   
26           1  Chocolate Sandwich Cookies        61             19     5.8   
27           1  Chocolate Sandwich Cookies        61             19     5.8   
28           1  Chocolate Sandwich Cookies        61             19     5.8   
29           1  Chocolate Sandwich Cookies        61             19     5.8   
30           1  Chocolate Sandwich Cookies        61             19     5.8   
31           1  Chocolate Sandwich Cookies        61             19     5.8   
32           1  Chocolate Sandwich Cookies        61             19     5.8   
33           1  Chocolate Sandwich Cookies        61             19     5.8   
34           1  Chocolate Sandwich Cookies        61             19     5.8   
35           1  Chocolate Sandwich Cookies        61             19     5.8   
36           1  Chocolate Sandwich Cookies        61             19     5.8   
37           1  Chocolate Sandwich Cookies        61             19     5.8   
38           1  Chocolate Sandwich Cookies        61             19     5.8   
39           1  Chocolate Sandwich Cookies        61             19     5.8   
40           1  Chocolate Sandwich Cookies        61             19     5.8   
41           1  Chocolate Sandwich Cookies        61             19     5.8   
42           1  Chocolate Sandwich Cookies        61             19     5.8   
43           1  Chocolate Sandwich Cookies        61             19     5.8   
44           1  Chocolate Sandwich Cookies        61             19     5.8   
45           1  Chocolate Sandwich Cookies        61             19     5.8   
46           1  Chocolate Sandwich Cookies        61             19     5.8   
47           1  Chocolate Sandwich Cookies        61             19     5.8   
48           1  Chocolate Sandwich Cookies        61             19     5.8   
49           1  Chocolate Sandwich Cookies        61  

In [33]:
# Derive order_frequency_flag columns using the loc() function

ords_prods_merge.loc[ords_prods_merge['median_days'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

In [34]:
ords_prods_merge.loc[(ords_prods_merge['median_days'] <= 20) & (ords_prods_merge['median_days'] > 10), 'order_frequency_flag'] = 'Regular customer'

In [37]:
ords_prods_merge.loc[ords_prods_merge['median_days'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [38]:
ords_prods_merge.head(100)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   
15           1  Chocolate Sandwich Cookies        61             19     5.8   
16           1  Chocolate Sandwich Cookies        61             19     5.8   
17           1  Chocolate Sandwich Cookies        61             19     5.8   
18           1  Chocolate Sandwich Cookies        61             19     5.8   
19           1  Chocolate Sandwich Cookies        61             19     5.8   
20           1  Chocolate Sandwich Cookies        61             19     5.8   
21           1  Chocolate Sandwich Cookies        61             19     5.8   
22           1  Chocolate Sandwich Cookies        61             19     5.8   
23           1  Chocolate Sandwich Cookies        61             19     5.8   
24           1  Chocolate Sandwich Cookies        61             19     5.8   
25           1  Chocolate Sandwich Cookies        61             19     5.8   
26           1  Chocolate Sandwich Cookies        61             19     5.8   
27           1  Chocolate Sandwich Cookies        61             19     5.8   
28           1  Chocolate Sandwich Cookies        61             19     5.8   
29           1  Chocolate Sandwich Cookies        61             19     5.8   
30           1  Chocolate Sandwich Cookies        61             19     5.8   
31           1  Chocolate Sandwich Cookies        61             19     5.8   
32           1  Chocolate Sandwich Cookies        61             19     5.8   
33           1  Chocolate Sandwich Cookies        61             19     5.8   
34           1  Chocolate Sandwich Cookies        61             19     5.8   
35           1  Chocolate Sandwich Cookies        61             19     5.8   
36           1  Chocolate Sandwich Cookies        61             19     5.8   
37           1  Chocolate Sandwich Cookies        61             19     5.8   
38           1  Chocolate Sandwich Cookies        61             19     5.8   
39           1  Chocolate Sandwich Cookies        61             19     5.8   
40           1  Chocolate Sandwich Cookies        61             19     5.8   
41           1  Chocolate Sandwich Cookies        61             19     5.8   
42           1  Chocolate Sandwich Cookies        61             19     5.8   
43           1  Chocolate Sandwich Cookies        61             19     5.8   
44           1  Chocolate Sandwich Cookies        61             19     5.8   
45           1  Chocolate Sandwich Cookies        61             19     5.8   
46           1  Chocolate Sandwich Cookies        61             19     5.8   
47           1  Chocolate Sandwich Cookies        61             19     5.8   
48           1  Chocolate Sandwich Cookies        61             19     5.8   
49           1  Chocolate Sandwich Cookies        61  

In [39]:
# Check the frequency of the order_frequency_flag column

ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: order_frequency_flag, dtype: int64

In [40]:
# Create a subset to view the missing values

df_nan = ords_prods_merge[ords_prods_merge['order_frequency_flag'].isnull() == True]

In [41]:
df_nan

product_id  \
6234909        10749   
12947653       21334   
13839012       22198   
14758536       23695   
21673807       33401   

                                                           product_name  \
6234909                                         Organic Red Bell Pepper   
12947653                                          Organic Peeled Garlic   
13839012  4X Ultra Concentrated Natural Laundry Detergent, Free & Clear   
14758536                                       California Veggie Burger   
21673807                                           Goat Cheese Crumbles   

          aisle_id  department_id  prices  order_id  user_id eval_set  \
6234909         83              4     5.7    895835   159838    prior   
12947653       123              4    10.2    895835   159838    prior   
13839012        75             17     1.7    895835   159838    prior   
14758536        42              1     4.7    895835   159838    prior   
21673807        21             16    14.8    895835   159838    prior   

          order_number  orders_day_of_week  order_hour_of_day  \
6234909              1                   0                 17   
12947653             1                   0                 17   
13839012             1                   0                 17   
14758536             1                   0                 17   
21673807             1                   0                 17   

          days_since_prior_order  add_to_cart_order  reordered _merge  \
6234909                      NaN                  3          0   both   
12947653                     NaN                  5          0   both   
13839012                     NaN                  1          0   both   
14758536                     NaN                  2          0   both   
21673807                     NaN                  6          0   both   

            price_range_loc  busiest_day  busiest_days busiest_period_of_day  \
6234909   Mid-range product  Busiest day  Busiest days        Average orders   
12947653  Mid-range product  Busiest day  Busiest days        Average orders   
13839012  Low-range product  Busiest day  Busiest days        Average orders   
14758536  Low-range product  Busiest day  Busiest days        Average orders   
21673807  Mid-range product  Busiest day  Busiest days        Average orders   

          max_order  loyalty_flag  average_spending spending_flag  \
6234909           1  New customer              7.42   Low spender   
12947653          1  New customer              7.42   Low spender   
13839012          1  New customer              7.42   Low spender   
14758536          1  New customer              7.42   Low spender   
21673807          1  New customer              7.42   Low spender   

          median_days order_frequency_flag  
6234909           NaN                  NaN  
12947653          NaN                  NaN  
13839012          NaN                  NaN  
14758536          NaN                  NaN  
21673807          NaN                  NaN

# 08. Exporting

In [43]:
# Export the updated dataframe to orders_products_merged.pkl

ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged.pkl'))